src/NodeApi/JSReference.cs

```diff
@@ -188,19 +237,18 @@ protected virtual void Dispose(bool disposing)
         if (!IsDisposed)
         {
             IsDisposed = true;
-            napi_ref handle = _handle; // To capture in lambda
 
             // The context may be null if the reference was created from a "no-context" scope such
             // as the native host. In that case the reference must be disposed from the JS thread.
-            if (SynchronizationContext == null)
+            if (_context == null)
```

Comment:
>_context == null [](http://example.com/codeflow?start=16&length=16)

The `_context` being null is a bit "sketchy". Ideally, `JSReference` must always be associated with a non-null object that wraps up `napi_env`. We must not use or delete the `napi_ref` if that `napi_env` wrapper is disposed.
When `napi_env` is deleted, it also deletes all associated `napi_ref` internally. Any `napi_ref` in `JSReference` must become invalid because they are pointers to deleted data. Calls to `napi_delete_reference` or any other `napi_ref` functions may cause a crash or memory corruption because they would use deleted `napi_env` and/or `napi_ref`.
The `napi_env` disposable wrapper must be for the `JSReference` to be the same as `JSValueScope` for `JSValue` - to protect wrapped handle. It must protect `napi_ref`, while `JSValueScope` protects `napi_value`.
The `napi_env` wrapper must be invalidated in response to finalizer passed to `napi_set_instance_data`.

Reply:
The reason for the `NoContext` scope type is that the `NativeHost` must not set the instance data, because the `ManagedHost` must set instance data to a managed `JSRuntimeContext` instance. User code should never encounter this case because this is a special case for the native host, as commented there.

Currently the native host only creates one reference and keeps it for the process lifetime, so it never actually disposes the reference. So this code never executes when `_context` is null. But I wasn't sure whether the native would ever need to do more things with references in the future. If you prefer, we could change this so that the native host _leaks_ references instead of disposing them, which would be fine if the only reference(s) it uses are process lifetime.

Reply:
> Ideally, JSReference must always be associated with a non-null object that wraps up `napi_env`.

Currently we wrap the `napi_env` in `JSValueScope`, but that doesn't work for references since they may outlive a scope. I keep thinking maybe we should create a `JSEnvironment` class to wrap the `napi_env`. Then, `JSValueScope`, `JSReference`, and `JSRuntimeContext` could all reference a `JSEnvironment` instance. But I hesitate to create yet another class for maintaining context.

Reply:
Could we use the `JSRuntimeContext` for that?

Reply:
I think this is good enough for now, with the added check I added to validate the `_env` matches the current thread-local environment. But I will re-evaluate after the other planned changes to how we track the environment context.